File: Problems/0396-rotate-function.py
class Solution(object):
    def maxRotateFunction(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        n = len(nums)
        total = sum(nums)
        cur = 0

        for i, x in enumerate(nums):
            cur += i * x

        ans = cur

        for i in range(n - 1, 0, -1):
            cur += total - n * nums[i]
            if cur > ans:
                ans = cur

        return ans

File: Problems/0396-rotate-function.cpp
#include <bits/stdc++.h>
using namespace std;

// LeetCode solution starts from here

class Solution {
public:
    int maxRotateFunction(vector<int>& nums) {
        int n = nums.size();
        long long total = 0, cur = 0;

        for (int i = 0; i < n; i++) {
            total += nums[i];
            cur += 1LL * i * nums[i];
        }

        long long ans = cur;

        for (int i = n - 1; i > 0; i--) {
            cur += total - 1LL * n * nums[i];
            ans = max(ans, cur);
        }

        return ans;
    }
};
